### PR TITLE
Support to reload endpoints

### PR DESCRIPTION
**Pull request by awschristou**
## Description
This change adds a `Reload` method to RegionEndpoint, allowing callers to provide updated endpoints manifest data. The updated manifest data is used by the SDK when instantiating new service clients.

I had to update some code around the SDK that made assumptions around RegionEndpoint object equality with static fields like RegionEndpoint.USEast1. These checks now test against the region system name, treating this as a region endpoint's primary ID. The suppressions for analyzer CR1004 (Do not use static readonly RegionEndpoint members from within the SDK) were used to identify possible places in the SDK to change.

Notes:
* RegionEndpoint has static readonly fields like RegionEndpoint.USEast1. When Reload is called, these instances are not re-created, and properties on these objects do not get updated.
* Equality tests comparing RegionEndpoint static fields to RegionEndpoint objects retieved with GetEndpoint are not guaranteed to pass after calling Reload

## Motivation and Context
This change allows applications like the AWS Toolkit to obtain updated endpoints.json manifest files and leverage them without requiring (1) application updates (2) SDK package reference updates. This also avoids the need to duplicate the SDK's logic for endpoint resolution into any application (like the Toolkit).

## Testing

* provided unit test to verify that Reload updates the region endpoint details
* locally ran a test to instantiate a service client after calling Reload. The service client contained config details relating to the updated region endpoints.,

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/src/Core/RegionEndpoint/RegionEndpoint.cs [hunk 1]
```diff
@@ -207,7 +213,81 @@ public static RegionEndpoint GetBySystemName(string systemName)
         /// <returns></returns>
         public static RegionEndpoint GetRegionEndpointOverride(RegionEndpoint regionEndpoint)
         {
-            return _hashRegionEndpointOverride.ContainsKey(regionEndpoint) ? _hashRegionEndpointOverride[regionEndpoint] : null;
+            lock (_hashRegionEndpointOverride)
```
Comment (normj) [inline sdk/src/Core/RegionEndpoint/RegionEndpoint.cs:216]: We should use a `ReaderWriterLockSlim` lock so we are obtaining an exclusive lock for every SigV4 sign request. Here is a recent example in the SDK using the `ReaderWriterLockSlim`. https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Credentials/DefaultInstanceProfileAWSCredentials.cs#L33

**Review by normj — CHANGES_REQUESTED** [review]

File: sdk/test/UnitTests/Custom/Runtime/RegionEndpointTests.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                return JsonMapper.ToObject(reader);
+            }
+        }
+
+        private Stream GetStream()
+        {
+            string assemblyLocation = typeof(RegionEndpointProviderV3).Assembly.Location;
+            if (!string.IsNullOrEmpty(assemblyLocation))
+            {
+                string endpointsPath = @"C:\codebase\v3\aws-sdk-net\sdk\src\Core\endpoints.json";
```
Comment (normj) [inline sdk/test/UnitTests/Custom/Runtime/RegionEndpointTests.cs:79]: This path is hardcoded to where you check out the repo. You need to do something using a relative path from this project.

**Review by normj — CHANGES_REQUESTED** [review]

File: sdk/test/UnitTests/Custom/Runtime/RegionEndpointTests.cs (hunk 2 above)
Comment (awschristou) [reply sdk/test/UnitTests/Custom/Runtime/RegionEndpointTests.cs:79]: How embarrassing - I forgot to circle back to this before wrapping up! I'll get that fixed up.

Comment (awschristou) [reply sdk/test/UnitTests/Custom/Runtime/RegionEndpointTests.cs]: Fixed

**Review by normj — APPROVED** [review]

File: sdk/src/Services/SecurityToken/Custom/AmazonSecurityTokenServiceConfig.Extension.cs [hunk 3]
```diff
@@ -100,6 +100,16 @@ public partial class AmazonSecurityTokenServiceConfig : ClientConfig
          };
 #endif
 
+        private static readonly ISet<string> legacyGlobalRegionSystemNames = new HashSet<string>();
```
Comment (ashovlin) [inline sdk/src/Services/SecurityToken/Custom/AmazonSecurityTokenServiceConfig.Extension.cs:103]: `ISet` doesn't extend back to .NET Framework 3.5

https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iset-1?view=net-5.0#applies-to

This command failed during the dry-run:
```
  Custom\AmazonSecurityTokenServiceConfig.Extension.cs(103,33): error CS0246: The type or namespace name 'ISet<>' could not be found (are you missing a using directive or an assembly reference?) [D:\JenkinsWorkspaces\Stage-release-dry-run\AWSDotNetPublic\sdk\src\Services\SecurityToken\AWSSDK.SecurityToken.Net35.csproj] [D:\JenkinsWorkspaces\Stage-release-dry-run\AWSDotNetPublic\buildtools\build.proj]

  D:\JenkinsWorkspaces\Stage-release-dry-run\AWSDotNetPublic\buildtools\build.proj(172,9): error MSB3073: The command "dotnet build -c Release --no-incremental AWSSDK.Net35.sln /p:AWSKeyFile=c:\aws\AWSDotNet_Private.snk;SignAssembly=true;RunAnalyzersDuringBuild=true;RuleSetFileForBuild=" exited with code 1.
```

**Review by ashovlin — CHANGES_REQUESTED** [review]

File: sdk/src/Services/SecurityToken/Custom/AmazonSecurityTokenServiceConfig.Extension.cs (hunk 3 above)
Comment (awschristou) [reply sdk/src/Services/SecurityToken/Custom/AmazonSecurityTokenServiceConfig.Extension.cs:102]: I've revised it to follow the existing ifdef convention

**Review by boblodgett — APPROVED** [review]

**Review by ashovlin — APPROVED** [review]

**Pull request merged**